IPython/core/application.py

```diff
@@ -215,7 +215,7 @@ def excepthook(self, etype, evalue, tb):
             return crashhandler.crash_handler_lite(etype, evalue, tb)
     
     def _ipython_dir_changed(self, name, old, new):
-        if old is not None:
+        if old is not None and old is not Undefined:
```

Comment:
In traitlets.py Line 449, I changed the value of `old` from `None` to `Undefined` in case of key error. 

The reason for that change, is that to roll back in case of bulk rejection, I need to distinguish between unset attributes and attributes set to None.